src/rt/sched/cown.h

```diff
@@ -525,29 +527,66 @@ namespace verona::rt
       const auto last = body->count - 1;
       assert(body->index <= last);
 
-      const auto high_priority = behaviour_requires_high_priority(body);
+      auto high_priority = false;
+      if (body->index == 0)
+      {
+        // If priority is needed for any cown in this message, start unmuting
+        // cowns in the body so that they can start running messages in their
+        // queue.
+        high_priority = std::any_of(
+          &body->cowns[0], &body->cowns[body->count], [](const auto* c) {
+            return (
+              c->bp_state.load(std::memory_order_acquire) & PriorityMask::High);
+          });
+      }
+
       for (; body->index < body->count; body->index++)
       {
-        MultiMessage* m = MultiMessage::make_message(alloc, body, epoch);
+        auto m = MultiMessage::make_message(alloc, body, epoch);
         auto* next = body->cowns[body->index];
         Systematic::cout() << "MultiMessage " << m << ": fast requesting "
                            << next << ", index " << body->index << std::endl;
 
+        if (body->index > 0)
         {
-          // Hold epoch in case priority needs to be raised after message is
-          // placed in queue. TODO: only hold if `high_priority` is true
-          Epoch e(alloc);
+          // Double check the priority of the most recently acquired cown to
+          // prevent deadlock.
+          auto* cur = body->cowns[body->index - 1];
+          high_priority = high_priority ||
+            (cur->bp_state.load(std::memory_order_acquire) & PriorityMask::High)
+#ifdef USE_SYSTEMATIC_TESTING
+            || Systematic::coin(3)
+#endif
+            ;
+          yield();
```

Comment:
I am not happy this is correct.  I understand what you are trying to do, but I think the issue is that a thread can set a cown to high-priority and not see a `blocker`, but the owning scheduler thread does not see it is set to high-priority, so does not perform `backpressure_unblock(next, e);` on the subsequent cowns.

I am coming to the conclusion that compressing `bp_state` and `blocker` into a single machine word is pretty much the only solution. Or you need to add fences.

Reply:
I think something like
```
when () {
    for (lots of times) { 
        when (c_muted1) { 
            when (c_overloaded1) { 
                something slow 
            }
        }
    }
}

when () {
    for (lots of times) { 
        when (c_muted2) { 
            when (c_overloaded2) { 
                something slow 
            }
        }
    }
}

when (c_overloaded1, c_muted2)
{

}
```
might exhibit the problem.  `backpressure3` almost does this, but the pair of the muted cown and overloaded cown are too related.  The muting is caused by the overloading, which induces additional orderings.  The example above has independence between the muted and the overloaded cown.

Reply:
The priority and blocker have been merged into the same atomic variable. The test you have described has also been added to `func-backpressure`.